### PR TITLE
Implement parametric timestamp with time zone

### PR DESCRIPTION
**Pull request by martint**

### CODE REVIEW

**Review by dain — COMMENTED** [review]
Some comments so far... these are on the out dated diff, and may not apply anymore

File: presto-spi/src/main/java/io/prestosql/spi/type/LongTimestampWithTimeZone.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        return picosOfMilli;
+    }
+
+    public short getTimeZoneKey()
+    {
+        return timeZoneKey;
+    }
+
+    @Override
+    public boolean equals(Object o)
```
[the remark this answers is not in the document]
Comment (martint) [reply presto-spi/src/main/java/io/prestosql/spi/type/LongTimestampWithTimeZone.java:73]: Good point. I don't think we need them, but let me check.

**Review by dain — APPROVED** [review]
Couple of minor comments but otherwise looks good

File: presto-main/src/main/java/io/prestosql/operator/scalar/timestamptz/DateTrunc.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        return updateMillisUtc(epochMillis, packedEpochMillis);
+    }
+
+    @LiteralParameters({"x", "p"})
+    @SqlType("timestamp(p) with time zone")
+    public static LongTimestampWithTimeZone truncate(
+            @SqlType("varchar(x)") Slice unit,
+            @SqlType("timestamp(p) with time zone") LongTimestampWithTimeZone timestamp)
+    {
+        ISOChronology chronology = getChronology(getTimeZoneKey(timestamp.getTimeZoneKey()));
```
[the remark this answers is not in the document]
Comment (martint) [reply presto-main/src/main/java/io/prestosql/operator/scalar/timestamptz/DateTrunc.java:57]: The core implementation is just one relatively simple line of code. It doesn't seem worth it.

**Pull request merged**